src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -221,11 +221,12 @@ private BoundExpression ApplyConvertedTypes(BoundExpression expr, TupleBinaryOpe
             if (leftCardinality != rightCardinality)
             {
                 Error(diagnostics, ErrorCode.ERR_TupleSizesMismatchForBinOps, node, leftCardinality, rightCardinality);
-                return new TupleBinaryOperatorInfo.Multiple(ImmutableArray<TupleBinaryOperatorInfo>.Empty, leftConvertedTypeOpt: null, rightConvertedTypeOpt: null);
+                return TupleBinaryOperatorInfo.Multiple.ErrorInstance;
             }
 
-            ImmutableArray<BoundExpression> leftParts = GetTupleArgumentsOrPlaceholders(left);
-            ImmutableArray<BoundExpression> rightParts = GetTupleArgumentsOrPlaceholders(right);
+            var (leftParts, leftNames) = GetTupleArgumentsOrPlaceholders(left);
+            var (rightParts, rightNames) = GetTupleArgumentsOrPlaceholders(right);
+            ReportNamesMismatchesIfAny(left, right, diagnostics);
```

Comment:
>ReportNamesMismatchesIfAny(left, right, diagnostics); [](start = 12, length = 53)

We already calculated names and call function that will do this again in a different way. Can we reuse the information?

Reply:
We could, but it makes the code more complex and I don't think it helps performance (getting the same array a second time should be cheap).

---
In reply to: [173944175](https://github.com/dotnet/roslyn/pull/25396#discussion_r173944175) [](ancestors = 173944175)

Reply:
>  it makes the code more complex

It can actually make code much simpler because we might be able to get rid of the getDetails helper completely. The tuple literal specific checks can simply be inlined, instead of producing tuple with defaults for non-tuple cases.

---
In reply to: [173945636](https://github.com/dotnet/roslyn/pull/25396#discussion_r173945636) [](ancestors = 173945636,173944175)